Adding more unit tests

rmkraus opened the pull request:
The Joker has Batman. Lex Luthor has Superman. Dr. Claw has Inspector Gadget. All evil demands good. It is the Yin and Yang that drives our crazy existence on this blue marble.

I have racked up some bad karma lately (#1012) and need to repent. This is my repentance. More unit tests.

Don't merge yet, I'm opening the PR early so I can track my progress.

rmkraus committed:
Added tests for package utilities

rmkraus committed:
Added tests to check setup and config of universal media player.

rmkraus committed:
Fixed typo in universal media player test.

rmkraus committed:
Removed unused private method from universal media player

The universal media player contained a private method that was replaced
by the update method. It was meant to be removed and wasn’t. This
commit removed that method.

rmkraus committed:
Added test for universal mp service routing.

Added tests to ensure that the Universal Media Player is routing
service calls correctly.

rmkraus committed:
Added test for logger component.

balloob commented on tests/util/test_package.py:
```diff
@@ -0,0 +1,58 @@
+
+        self.assertTrue(package.install_package(TEST_EXIST_REQ))
+
+    def test_install_package_locally(self):
```
I'm not too sure about these tests. It's good to test it but adding a dependency on external resources makes it slow, unable to test without internet etc. What about you only test it with either a local cached copy of the `helloworld` package (using Betamax, see [YR tests](https://github.com/balloob/home-assistant/blob/dev/tests/components/sensor/test_yr.py) for an example) or install a local zipfile.

rmkraus replied:
I'll install a local zip. I'm a bit out of my element with Betamax, but I'm not sure we can use Betamax because the download is running in a subprocess.

balloob commented on tests/util/test_package.py:
```diff
@@ -0,0 +1,58 @@
+            TEST_NEW_REQ, True, self.lib_dir.name))
+
+        sys.path.insert(0, self.lib_dir.name)
+        import pyhelloworld3
```
```python
try:
    import pyhelloworld3
except ImportError:
    self.fail('Unable to import pyhelloworld3 after installing it.')
```

balloob commented on tests/util/test_package.py:
```diff
@@ -0,0 +1,58 @@
+        self.assertTrue(package.install_package(
+            TEST_NEW_REQ, True, self.lib_dir.name))
+
+        sys.path.insert(0, self.lib_dir.name)
```
Can you use the `bootstrap.mount_local_lib_path(config_dir)` helper method. Yes, it does the same but it also means we keep doing the same if we ever update that method.

balloob commented:
+1% test coverage, very very nice :+1:

rmkraus committed:
Additional testing for logger component

Added an additional test for the logger component the validates the
filtering logic of the filters that were created during setup.

rmkraus committed:
Revised package util tests

The package util tests were revised to pull the external library
pyhelloworld3 from an internal source rather than external. This speeds
up tests, makes tests more reliable, and removes dependency on internet
connection.

balloob commented on tests/components/media_player/test_universal.py:
```diff
@@ -340,3 +404,38 @@ def test_supported_media_commands_children_and_cmds(self):
+        ump.update()
+
+        ump.turn_off()
+        self.assertEqual(len(self.mock_mp_2.turn_off_service_calls), 1)
```
You want to put the expected value first.

balloob commented on tests/util/test_package.py:
```diff
@@ -0,0 +1,61 @@
+
+    def tearDown(self):
+        """ Remove local library """
+        del self.tmp_dir
```
Use [`cleanup`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory):
```python
self.tmp_dir.cleanup()
```

rmkraus committed:
Cleaned up universal MP tests

Changed all assertEquals in universal media player tests to have
assumed value first.

rmkraus committed:
Better tear down of util/package tests

Explicitly removed temp directory at the end of util/package unit tests.

rmkraus commented:
I really underestimated how many checks it takes to raise coverage. My initial goal was to hit around 90%, but that seems lofty now.

balloob commented:
Test coverage is only an indicator if your test coverage is bad. It only measures if a line is hit, not how many code paths are explored. Once you start adding tests for 1 file, just importing it adds like 30% for free because it reads the class definition. But making sure every method gets called is the tricky part.

Coveralls gives a good indication where we lack testing. With this PR we would have 6302 of 7248 relevant lines covered. To get to 90% we would need another 230 lines covered. Since our files are pretty small, that's a pretty big effort

balloob commented:
This all looks good so you can merge when you're ready

rmkraus committed:
Added test for entity customization

Added test for entity customization from configuration. Processes a
sample configuration to hide an entity, creates the entity, updates ha
state, and then verifies customization made it through.

rmkraus committed:
Added tests for command sensor

Added tests to create and check basic functionality of command sensor.

rmkraus committed:
Additional tests for Command Sensor.

1. Moved template testing out of main test.
2. Added test for bad command.

rmkraus committed:
Another test for for command sensor

Added a test for command sensors with bad configurations.

rmkraus committed:
Added tests for Binary Command Sensor

rmkraus committed:
More tests for Binary Command Sensor

1. Added a test for detecting STATE_OFF
2. Fixed tests for detecting STATE_ON

rmkraus committed:
Added test for Introduction component

This test may seem useless, but it is good to ensure that default
components don’t ever crash HASS.

rmkraus commented:
2.6% is good enough for now, I suppose. You want to check everything over one more time?

balloob commented:
You must have felt very guilty :yum:
